walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -1,23 +1,68 @@
-// Copyright 2019 Google LLC
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     https://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+
+// TIME COMPLEXITY: O(E * R), where E is the number of events and R is the number of members in the request
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        if (request.getDuration() > 1440)
+            return new ArrayList<>();
+        else if (events == null || events.isEmpty())
+            return new ArrayList<>(Collections.singletonList(TimeRange.fromStartEnd(0, 1440, false)));
+        else{
+            List<List<Integer>> meetingTimeRanges = new ArrayList<>();
+
+            for (Event singleEvent: events){
+                if (requestedMemberInExistingMeeting(request.getAttendees(), singleEvent.getAttendees()))
+                    meetingTimeRanges.add(Arrays.asList(singleEvent.getWhen().start(), singleEvent.getWhen().end()));
+            }
+
+            int i = 0;
+            while (i < meetingTimeRanges.size()-1){
+                if (overlaps(meetingTimeRanges.get(i), meetingTimeRanges.get(i+1)))
+                    meetingTimeRanges.add(i, joinOverlappingTimeRanges(meetingTimeRanges.remove(i), meetingTimeRanges.remove(i)));
+                else
+                    i++;
+            }
+
+            ArrayList<TimeRange> freeTimeSlots = new ArrayList<>();
+
+            int startTime = 0;
+            int endTime;
+            for (List<Integer> singleMeetingTime: meetingTimeRanges){
+                endTime = singleMeetingTime.get(0);
+                if (endTime - startTime >= request.getDuration())
+                    freeTimeSlots.add(TimeRange.fromStartEnd(startTime, endTime, false));
+                startTime = singleMeetingTime.get(1);
+            }
+            if (startTime < 1440)
+                freeTimeSlots.add(TimeRange.fromStartEnd(startTime, 1440, false));
+
+            return freeTimeSlots;
+        }
+    }
+    public static boolean requestedMemberInExistingMeeting(Collection<String> requestedMembers, Set<String> eventMembers){
```

Comment:
Can you do this more simply using existing Java methods on collections? There should be a method that returns the intersection of two collections.

Reply:
Yes, I found that the Collections.disjoint() method does exactly what my requestedMemberInExistingMeeting() method does. Thank you.